src/platform_impl/linux/wayland/seat/keyboard/handlers.rs

```diff
@@ -0,0 +1,151 @@
+            };
+
+            let virtual_keycode = keymap::keysym_to_vkey(keysym);
+
+            event_sink.push_window_event(
+                #[allow(deprecated)]
+                WindowEvent::KeyboardInput {
+                    device_id: crate::event::DeviceId(crate::platform_impl::DeviceId::Wayland(
+                        DeviceId,
+                    )),
+                    input: KeyboardInput {
+                        state,
+                        scancode: rawkey,
+                        virtual_keycode,
+                        modifiers: *inner.modifiers_state.borrow(),
+                    },
+                    is_synthetic: false,
+                },
+                window_id,
+            );
+
+            // Send ReceivedCharacter event only on ElementState::Pressed.
+            if ElementState::Released == state {
+                return;
+            }
+
+            if let Some(txt) = utf8 {
+                for ch in txt.chars() {
+                    event_sink.push_window_event(WindowEvent::ReceivedCharacter(ch), window_id);
+                }
+            }
+        }
+        KeyboardEvent::Repeat {
+            rawkey,
+            keysym,
+            utf8,
+            ..
+        } => {
+            let window_id = match inner.target_window_id {
+                Some(window_id) => window_id,
+                None => return,
+            };
+
+            let virtual_keycode = keymap::keysym_to_vkey(keysym);
+
+            event_sink.push_window_event(
+                #[allow(deprecated)]
+                WindowEvent::KeyboardInput {
+                    device_id: crate::event::DeviceId(crate::platform_impl::DeviceId::Wayland(
+                        DeviceId,
+                    )),
+                    input: KeyboardInput {
+                        state: ElementState::Pressed,
+                        scancode: rawkey,
+                        virtual_keycode,
+                        modifiers: *inner.modifiers_state.borrow(),
+                    },
+                    is_synthetic: false,
+                },
+                window_id,
+            );
+
+            if let Some(txt) = utf8 {
+                for ch in txt.chars() {
+                    event_sink.push_window_event(WindowEvent::ReceivedCharacter(ch), window_id);
+                }
+            }
+        }
+        KeyboardEvent::Modifiers { modifiers } => {
+            let modifiers = ModifiersState::from_wayland(modifiers);
+            if let Some(window_id) = inner.target_window_id {
+                *inner.modifiers_state.borrow_mut() = modifiers;
+
+                event_sink.push_window_event(WindowEvent::ModifiersChanged(modifiers), window_id);
+            } else {
+                // Compositor must send modifiers after wl_keyboard::enter, however certain
+                // compositors are still sending it before, so stash such events and send
+                // them on wl_keyboard::enter.
```

Comment:
Yes, because winit sends events downstream once it gained keyboard focus, however some compositors send modifiers(which is a compositor bug) before keyboard focus, so I've added a work around for those, since winit will ignore that even in general, since no window is being focused.

Reply:
Ah, so the problem is not about event order of key/modifier event, but modifier/focus event. That makes sense, yeah.

Though generally do we need to filter events on Wayland to require focus? Isn't this basically done automatically for us? Or is it filtered in some cross-platform place?

Reply:
>Though generally do we need to filter events on Wayland to require focus? Isn't this basically done automatically for us? Or is it filtered in some cross-platform place?

It's not filtering entirely, it's targeting the right winit window since you have one keyboard for all the windows out there. Filtering is done naturally, since we don't know what window we're going to focus without having keyboard focus. The thing is that some compositors are sending keyboard events before keyboard focus, and that code is just to work around buggy compositors like GNOME, since it only fails with modifiers, and send those before any keyboard focus, so winit don't know to which window we're sending events.